### PR TITLE
Use gox for building

### PR DESCRIPTION
**Pull request by sethvargo**

### CODE REVIEW

File: scripts/dist.sh [hunk 1]
```diff
@@ -22,53 +16,45 @@ DIR="$( cd -P "$( dirname "$SOURCE" )/.." && pwd )"
...
+# for ARCHIVE in ./pkg/dist/*; do
+#     ARCHIVE_NAME=$(basename ${ARCHIVE})
 
-    echo Uploading: $ARCHIVE_NAME
-    curl \
-        -T ${ARCHIVE} \
-        -umitchellh:${BINTRAY_API_KEY} \
-        "https://api.bintray.com/content/mitchellh/consul/consul/${VERSION}/${ARCHIVE_NAME}"
-done
+#     echo Uploading: $ARCHIVE_NAME
```
Comment (sethvargo) [inline scripts/dist.sh]: Leaving this commented out for now intentionally

File: Makefile [hunk 2]
```diff
@@ -2,11 +2,24 @@ DEPS = $(shell go list -f '{{range .TestImports}}{{.}} {{end}}' ./...)
...
 
 all: deps format
 	@mkdir -p bin/
 	@bash --norc -i ./scripts/build.sh
 
+# bin generates the releasable binaries
+bin: generate
+	@sh -c "'$(CURDIR)/scripts/build.sh'"
+
+# dev creates binares for testing locally. There are put into ./bin and $GOPATH
```
Comment (slackpad) [inline Makefile]: sp - "These"

Comment (ryanbreen) [inline Makefile]: sp - "binaries"

File: Makefile [hunk 3]
```diff
@@ -46,10 +65,15 @@ vet:
 		echo "and fix them if necessary before submitting the code for reviewal."; \
 	fi
 
+# generate runs `go generate` to build the dynamically generated source files
+generate: deps
```
Comment (slackpad) [inline Makefile]: Do we actually have these or is this just a standard thing?

File: scripts/build.sh [hunk 4]
```diff
@@ -16,37 +16,67 @@ GIT_COMMIT=$(git rev-parse HEAD)
...
+echo "==> Removing old directory..."
+rm -f bin/*
+rm -rf pkg/*
+mkdir -p bin/
 
-# On OSX, we need to use an older target to ensure binaries are
-# compatible with older linkers
-if [ "$(go env GOOS)" = "darwin" ]; then
-    export MACOSX_DEPLOYMENT_TARGET=10.6
+# If its dev mode, only build for ourself
```
Comment (slackpad) [inline scripts/build.sh]: it's :-)

File: scripts/build.sh [hunk 5]
```diff
@@ -16,37 +16,67 @@ GIT_COMMIT=$(git rev-parse HEAD)
...
+
+# Move all the compiled things to the $GOPATH/bin
+GOPATH=${GOPATH:-$(go env GOPATH)}
+case $(uname) in
+    CYGWIN*)
+        GOPATH="$(cygpath $GOPATH)"
+        ;;
+esac
+OLDIFS=$IFS
+IFS=: MAIN_GOPATH=($GOPATH)
```
Comment (slackpad) [inline scripts/build.sh]: What does this do?

File: scripts/dist.sh [hunk 6]
```diff
@@ -22,53 +16,45 @@ DIR="$( cd -P "$( dirname "$SOURCE" )/.." && pwd )"
...
+if [ -z $NOSIGN ]; then
+  echo "==> Signing..."
+  gpg --default-key 348FFC4C --detach-sig ./consul_${VERSION}_SHA256SUMS
+fi
 popd
 
-# Upload
-for ARCHIVE in ./dist/pkg/*; do
-    ARCHIVE_NAME=$(basename ${ARCHIVE})
+# # Upload
```
Comment (slackpad) [inline scripts/dist.sh]: Why not just delete this stuff?

**Comment by slackpad** [comment]
Just some little things and a few q-s.

File: Makefile (hunk 3 above)
Comment (sethvargo) [inline Makefile]: It's just a standard thing that all our projects do. It's super super fast.

File: scripts/build.sh (hunk 5 above)
Comment (sethvargo) [inline scripts/build.sh]: I'm not 100% sure, I copied it from Terraform. It has to do with putting the binary in the proper place on windows under cygwin.

**Comment by slackpad** [comment]
LGTM

**Pull request merged**